flare/gp_from_aimd.py

```diff
@@ -252,18 +252,357 @@ def __init__(
+        train_counter = 0  # Track # of times training done
+
+        # Keep track of which atoms trigger force / uncertainty condition
+        training_plan = {}
+
+        # MAIN LOOP - Frames
+        for i, cur_frame in enumerate(frames):
+            frame_start_time = time.time()
+            logger.info(f"=====NOW ON FRAME {i}=====")
+
+            # If no predict_atoms_per_element was specified, predict_atoms
+            # will be equal to every atom in the frame.
+            predict_atoms = subset_of_frame_by_element(cur_frame, predict_atoms_per_elt)
+
+            # Atoms which are skipped will have NaN as their force / std values
+            local_energies = None
+
+            # Three different predictions: Either MGP, GP with energy,
+            # or GP without
+            if self.mgp:
+                pred_forces, pred_stds, local_energies = self.pred_func(
+                    structure=cur_frame,
+                    mgp=self.gp,
+                    write_to_structure=False,
+                    selective_atoms=predict_atoms,
+                    skipped_atom_value=np.nan,
+                    energy=True,
+                )
+            elif self.calculate_energy:
+                pred_forces, pred_stds, local_energies = self.pred_func(
+                    structure=cur_frame,
+                    gp=self.gp,
+                    n_cpus=self.n_cpus,
+                    write_to_structure=False,
+                    selective_atoms=predict_atoms,
+                    skipped_atom_value=np.nan,
+                )
+            else:
+                pred_forces, pred_stds = self.pred_func(
+                    structure=cur_frame,
+                    gp=self.gp,
+                    n_cpus=self.n_cpus,
+                    write_to_structure=False,
+                    selective_atoms=predict_atoms,
+                    skipped_atom_value=np.nan,
+                )
+
+            # Get Error
+            dft_forces = cur_frame.forces
+            dft_energy = cur_frame.energy
+            error = np.abs(pred_forces - dft_forces)
+
+            # Create dummy frame with the predicted forces written
+            dummy_frame = deepcopy(cur_frame)
+            dummy_frame.forces = pred_forces
+            dummy_frame.stds = pred_stds
+            cur_frame.stds = pred_stds
+
+            self.output.write_gp_dft_comparison(
+                curr_step=i,
+                frame=dummy_frame,
+                start_time=frame_start_time,
+                dft_forces=dft_forces,
+                dft_energy=dft_energy,
+                error=error,
+                local_energies=local_energies,
+                KE=0,
+                cell=cur_frame.cell,
+            )
+
+            logger.debug(
+                f"Single frame calculation time {time.time()-frame_start_time}"
+            )
+
+            if i < train_frame:
+                # Noise hyperparameter & relative std tolerance is not for mgp.
```

Comment:
Might be better to wrap the code in this `if` condition into a function (e.g. `add_training_data` or so), so that this main loop looks clearer in terms of code structure

Reply:
I do like this idea, and appreciate the thoughtful feedback-- I think I will add it in a future refactoring.